From 31097e2b996ed463ca97d3df618a614c875386c5 Mon Sep 17 00:00:00 2001
From: Dan Streetman <ddstreet@canonical.com>
Date: Tue, 29 Jun 2021 09:13:22 -0400
Subject: [PATCH 1/3] time: simplify get_timezones()
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1858210
Origin: upstream, https://github.com/systemd/systemd/pull/20066

The function can be simplified by using extract_many_words() and strv_extend()
---
 src/basic/time-util.c | 56 +++++++++++++++----------------------------
 1 file changed, 19 insertions(+), 37 deletions(-)

--- a/src/basic/time-util.c
+++ b/src/basic/time-util.c
@@ -1205,24 +1205,14 @@ bool ntp_synced(void) {
 int get_timezones(char ***ret) {
         _cleanup_fclose_ FILE *f = NULL;
         _cleanup_strv_free_ char **zones = NULL;
-        size_t n_zones = 0, n_allocated = 0;
         int r;
 
         assert(ret);
 
-        zones = strv_new("UTC");
-        if (!zones)
-                return -ENOMEM;
-
-        n_allocated = 2;
-        n_zones = 1;
-
         f = fopen("/usr/share/zoneinfo/zone1970.tab", "re");
         if (f) {
                 for (;;) {
-                        _cleanup_free_ char *line = NULL;
-                        char *p, *w;
-                        size_t k;
+                        _cleanup_free_ char *line = NULL, *cc = NULL, *co = NULL, *tz = NULL;
 
                         r = read_line(f, LONG_LINE_MAX, &line);
                         if (r < 0)
@@ -1230,45 +1220,34 @@ int get_timezones(char ***ret) {
                         if (r == 0)
                                 break;
 
-                        p = strstrip(line);
+                        const char *p = line;
 
-                        if (isempty(p) || *p == '#')
+                        /* Line format is:
+                         * 'country codes' 'coordinates' 'timezone' 'comments' */
+                        r = extract_many_words(&p, NULL, 0, &cc, &co, &tz, NULL);
+                        if (r < 0)
                                 continue;
 
-                        /* Skip over country code */
-                        p += strcspn(p, WHITESPACE);
-                        p += strspn(p, WHITESPACE);
-
-                        /* Skip over coordinates */
-                        p += strcspn(p, WHITESPACE);
-                        p += strspn(p, WHITESPACE);
-
-                        /* Found timezone name */
-                        k = strcspn(p, WHITESPACE);
-                        if (k <= 0)
+                        /* Lines that start with # are comments. */
+                        if (*cc == '#')
                                 continue;
 
-                        w = strndup(p, k);
-                        if (!w)
-                                return -ENOMEM;
-
-                        if (!GREEDY_REALLOC(zones, n_allocated, n_zones + 2)) {
-                                free(w);
-                                return -ENOMEM;
-                        }
-
-                        zones[n_zones++] = w;
-                        zones[n_zones] = NULL;
+                        r = strv_extend(&zones, tz);
+                        if (r < 0)
+                                return r;
                 }
-
-                strv_sort(zones);
-                strv_uniq(zones);
-
         } else if (errno != ENOENT)
                 return -errno;
 
-        *ret = TAKE_PTR(zones);
+        /* Always include UTC */
+        r = strv_extend(&zones, "UTC");
+        if (r < 0)
+                return -ENOMEM;
+
+        strv_sort(zones);
+        strv_uniq(zones);
 
+        *ret = TAKE_PTR(zones);
         return 0;
 }
 
